File: 59.螺旋矩阵-ii.c
/*
 * @lc app=leetcode.cn id=59 lang=c
 *
 * [59] 螺旋矩阵 II
 */

// @lc code=start

/**
 * Return an array of arrays of size *returnSize.
 * The sizes of the arrays are returned as *returnColumnSizes array.
 * Note: Both returned array and *columnSizes array must be malloced, assume caller calls free().
 */
int **generateMatrix(int n, int *returnSize, int **returnColumnSizes)
{

    int i = 0;
    int left = 0;
    int right = n - 1;
    int up = 0;
    int down = n - 1;
    int idx = 1;

    int **res = (int **)malloc(sizeof(int *) * n);
    *returnColumnSizes = (int *)malloc(sizeof(int) * n);
    *returnSize = n;

    for (i = 0; i < n; i++)
    {
        res[i] = (int *)malloc(sizeof(int) * n);
        (*returnColumnSizes)[i] = n;
    }

    while (up <= down && left <= right)
    {
        for (i = left; i <= right; i++)
            res[up][i] = idx++;
        up++;

        for (i = up; i <= down; i++)
            res[i][right] = idx++;
        right--;

        for (i = right; i >= left && up <= down; i--)
            res[down][i] = idx++;
        down--;

        for (i = down; i >= up && left <= right; i--)
            res[i][left] = idx++;
        left++;
    }
    return res;

    // int i = 0;
    // int up = 0;
    // int down = n - 1;
    // int left = 0;
    // int right = n - 1;
    // int idx = 1;

    // int **res = (int **)malloc(sizeof(int *) * n);
    // *returnColumnSizes = (int *)malloc(sizeof(int) * n);
    // *returnSize = n;

    // for (i = 0; i < n; i++)
    // {
    //     res[i] = (int *)malloc(sizeof(int) * n);
    //     (*returnColumnSizes)[i] = n;
    // }

    // while (up <= down && left <= right)
    // {
    //     for (i = left; i <= right; i++)
    //         res[up][i] = idx++;
    //     up++;

    //     for (i = up; i <= down; i++)
    //         res[i][right] = idx++;
    //     right--;

    //     for (i = right; i >= left && up <= down; i--)
    //         res[down][i] = idx++;
    //     down--;

    //     for (i = down; i >= up && left <= right; i--)
    //         res[i][left] = idx++;
    //     left++;
    // }

    // return res;
}

// @lc code=end

File: 59.螺旋矩阵-ii.py
# @before-stub-for-debug-begin
from python3problem59 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode.cn id=59 lang=python3
#
# [59] 螺旋矩阵 II
#

# @lc code=start
import numpy as np
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        # res = [[0] * n for _ in range(n)]
        res = np.array(9).reshape(3, 3)
        left = 0
        right = n - 1
        up = 0
        down = n - 1
        idx = 1

        while left <= right and up <= down:
            i = left
            while i <= right:
                # res.append(idx)
                res[up][i] = idx
                idx += 1
                i += 1
            up += 1

            i = up
            while i <= down:
                # res.append(idx)
                res[i][right] = idx
                idx += 1
                i += 1
            right -= 1

            i = right
            while i >= left and up <= down:
                res[down][i] = idx
                idx += 1
                i -= 1
            down -= 1

            i = down
            while i >= up and left <= right:
                res[i][left] = idx
                idx += 1
                i -= 1
            left += 1

        return res   

# @lc code=end

